File: backend/app/modules/attempts/controller.py
from flask import Blueprint, jsonify, request

from app.modules.attempts.service import AttemptService
from app.routes.auth import login_required

attempts_bp = Blueprint("attempts", __name__, url_prefix="/api/attempts")
service = AttemptService()


@attempts_bp.get("")
@login_required()
def list_attempts(current_user):
    return jsonify({"items": service.list_attempts(request.args.to_dict())}), 200


@attempts_bp.post("")
@login_required()
def create_attempt(current_user):
    return jsonify(service.create_attempt(request.get_json(force=True), current_user["id"])), 201


@attempts_bp.patch("/<int:attempt_id>")
@login_required(required_profiles={"ADMIN"})
def update_attempt(current_user, attempt_id: int):
    return jsonify(service.update_attempt(attempt_id, request.get_json(force=True))), 200


@attempts_bp.delete("/<int:attempt_id>")
@login_required(required_profiles={"ADMIN"})
def delete_attempt(current_user, attempt_id: int):
    service.delete_attempt(attempt_id)
    return jsonify({"message": "Tentativa excluida com sucesso."}), 200

File: backend/src/modules/attempts/controller.ts
import type { Express } from "express";

import { asyncHandler, getQueryParams } from "../../core/http";
import { loginRequired } from "../../routes/auth";
import type { SerializedUser } from "../../types";
import { AttemptService } from "./service";

const service = new AttemptService();

export function registerAttemptsRoutes(app: Express) {
  app.get(
    "/api/attempts",
    loginRequired(),
    asyncHandler(async (req, res) => {
      res.status(200).json({ items: await service.listAttempts(getQueryParams(req.query)) });
    }),
  );

  app.post(
    "/api/attempts",
    loginRequired(),
    asyncHandler(async (req, res) => {
      const currentUser = res.locals.currentUser as SerializedUser;
      res.status(201).json(await service.createAttempt(req.body ?? {}, currentUser.id));
    }),
  );

  app.patch(
    "/api/attempts/:attemptId",
    loginRequired(new Set(["ADMIN"])),
    asyncHandler(async (req, res) => {
      res.status(200).json(await service.updateAttempt(Number(req.params.attemptId), req.body ?? {}));
    }),
  );

  app.delete(
    "/api/attempts/:attemptId",
    loginRequired(new Set(["ADMIN"])),
    asyncHandler(async (req, res) => {
      await service.deleteAttempt(Number(req.params.attemptId));
      res.status(200).json({ message: "Tentativa excluida com sucesso." });
    }),
  );
}
